### PR TITLE
Add support for AnyDB

### PR DESCRIPTION
**Pull request by KimForss**
### Problem
We need support other databases than Hana. Customers are today using 4 different database backends for SAP

DB2
Oracle
SQL Server
ASE

### Solution
Create a new module that deploys the infrastructure required to support the AnyDB workload. The current database node in the json is sufficient for the support so no new nodes are required.

The sizing of the environment (VM SKU, disks) will be based on the database size. The size is used as the lookup key against the disk configuration json (anydb_sizes)

"databases": [
{
"platform": "Oracle",
"db_version": "7.5.1",
"os": {
"source_image_id": "/subscriptions/xxx/resourceGroups/xxx/providers/Microsoft.Compute/images/xxx",
"//": "os_type is required if source_image_id is used. Possible value: Linux, Windows",
"os_type": "Linux"
},
"size": "1024",
...

The deployment logic will check if the desired database is one of the 4 supported types and only then will the anydb infrastructure be deployed.

### CODE REVIEW

File: deploy/terraform/modules/anydb_node/outputs.tf [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/outputs.tf:1]:
```suggestion
```

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    { high_availability = local.anydb_ha },
+    { authentication = local.authentication }
+  )
+
+  dbnodes = flatten([
+    [
+      for database in local.anydb-databases : [
+        for idx, dbnode in database.dbnodes : {
+          platform       = local.anydb_platform,
+          name           = upper(local.anydb_ostype) == "WINDOWS" ? format("%s_xdbw%02d-0", upper(local.anydb_sid), idx) : format("%s_xdbl%02d-0", upper(local.anydb_sid), idx),
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:96]: Sorry overlooked this earlier, the logic would be if name not provided, use naming convention:
```suggestion
          name           = try(dbnode.name, upper(local.anydb_ostype) == "WINDOWS" ? format("%s_xdbw%02d-0", upper(local.anydb_sid), idx) : format("%s_xdbl%02d-0", upper(local.anydb_sid), idx)),
```

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+          authentication = local.authentication
+          sid            = local.anydb_sid
+        }
+      ]
+    ],
+    [
+      for database in local.anydb-databases : [
+        for idx, dbnode in database.dbnodes : {
+          platform       = local.anydb_platform,
+          name           = upper(local.anydb_ostype) == "WINDOWS" ? format("%s_xdbw%02d-1", upper(local.anydb_sid), idx) : format("%s_xdbl%02d-1", upper(local.anydb_sid), idx),
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:109]: same as above.

File: deploy/terraform/modules/anydb_node/vm-anydb.tf [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    username   = local.authentication.username
+    public_key = file(var.sshkey.path_to_public_key)
+  }
+
+  boot_diagnostics {
+    storage_account_uri = var.storage-bootdiag.primary_blob_endpoint
+  }
+  tags = {
+    environment = "SAP"
+    SID         = local.anydb_sid
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/vm-anydb.tf:68]: To be consistent with windows vm block.
```suggestion
    role        = "db"
    SID         = upper(local.anydb_sid)
```

File: deploy/terraform/modules/anydb_node/variables_local.tf (hunk 2 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf:96]: fixed

File: deploy/terraform/modules/anydb_node/variables_local.tf (hunk 3 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf:109]: fixed

File: deploy/terraform/modules/anydb_node/vm-anydb.tf (hunk 4 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/vm-anydb.tf:68]: fixed

File: deploy/terraform/modules/anydb_node/outputs.tf (hunk 1 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/outputs.tf]: fixed

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+variable "storage-bootdiag" {
+  description = "Details of the boot diagnostics storage account"
+}
+
+variable "ppg" {
+  description = "Details of the proximity placement group"
+}
+
+#TODO move to central
+variable "region_mapping" {
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:18]: You can keep TODO comments, but please remove region related naming convention from this PR.
We do not want to grow this PR larger and larger.

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    { os = local.anydb_ostype },
+    { filesystem = local.anydb_fs },
+    { high_availability = local.anydb_ha },
+    { authentication = local.authentication }
+  )
+
+  dbnodes = flatten([
+    [
+      for database in local.anydb-databases : [
+        for idx, dbnode in database.dbnodes : {
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:134]: database.dbnodes is missing default.
Please check sample [here](https://github.com/Azure/sap-hana/blob/735b36397eef144574ae4aec77fc4f6a42b96055/deploy/terraform/modules/hdb_node/variables_local.tf#L95-L97).

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+          platform       = local.anydb_platform,
+          name           = "${local.anydb_nameprefix}-${upper(local.anydb_ostype) == "WINDOWS" ? format("%sxdb%02dw", lower(local.anydb_sid), 0) : format("%sxdb%02dl", lower(local.anydb_sid), 0)}",
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:136]: This change still does not consider situation when VM name is provided from `database.dbnodes`.
If the name not provided, we use default naming convention.

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+          authentication = local.authentication
+          sid            = local.anydb_sid
+        }
+      ]
+    ],
+    [
+      for database in local.anydb-databases : [
+        for idx, dbnode in database.dbnodes : {
+          platform       = local.anydb_platform,
+          name           = "${local.anydb_nameprefix}-${upper(local.anydb_ostype) == "WINDOWS" ? format("%sxdb%02dw", lower(local.anydb_sid), 1) : format("%sxdb%02dl", lower(local.anydb_sid), 1)}",
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:149]: Same as above.

File: deploy/terraform/modules/anydb_node/infrastructure.tf [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  network_interface_id    = azurerm_network_interface.anydb[count.index].id
+  ip_configuration_name   = azurerm_network_interface.anydb[count.index].ip_configuration[0].name
+  backend_address_pool_id = azurerm_lb_backend_address_pool.anydb[0].id
+}
+
+# AVAILABILITY SET ================================================================================================
+
+resource "azurerm_availability_set" "anydb" {
+  count                        = local.enable_deployment ? 1 : 0
+  name                         = format("%s-avset", upper(local.anydb_nameprefix))
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/infrastructure.tf:48]: With nameprefix change, now you lose the `SID` and `xdb` in naming for avset and lb.
image

Please revert the change to use nameprefix for this PR.

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+  # If custom image is used, we do not overwrite os reference with default value
+  anydb_custom_image = try(local.anydb.os.source_image_id, "") != "" ? true : false
+
+  anydb_os = {
+    "source_image_id" = local.anydb_custom_image ? local.anydb.os.source_image_id : ""
+    "publisher"       = try(local.anydb.os.publisher, local.anydb_custom_image ? "" : "")
+    "offer"           = try(local.anydb.os.offer, local.anydb_custom_image ? "" : "")
+    "sku"             = try(local.anydb.os.sku, local.anydb_custom_image ? "" : "")
+    "version"         = try(local.anydb.os.version, local.anydb_custom_image ? "" : "latest")
+  }
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:103]: market place image information should given default if custom image not used (example [here](https://github.com/Azure/sap-hana/blob/735b36397eef144574ae4aec77fc4f6a42b96055/deploy/terraform/modules/hdb_node/variables_local.tf#L66-L70)). Otherwise terraform fails when "os" is not given in the input.
```
Error: Error creating Linux Virtual Machine "nancyc-sid-name-01-or1xdb00l" (Resource Group "nancyc-sid-name-01"): compute.VirtualMachinesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameter" Message="The value of parameter imageReference.publisher is invalid." Target="imageReference.publisher"

  on deploy/terraform/modules/anydb_node/vm-anydb.tf line 21, in resource "azurerm_linux_virtual_machine" "dbserver":
  21: resource azurerm_linux_virtual_machine "dbserver" {
```
This may be tricky since different DB may require different marketplace image?
```suggestion
  anydb_os = {
    "source_image_id" = local.anydb_custom_image ? local.anydb.os.source_image_id : ""
    "publisher"       = try(local.anydb.os.publisher, local.anydb_custom_image ? "" : "<publisher>")
    "offer"           = try(local.anydb.os.offer, local.anydb_custom_image ? "" : "<offer>")
    "sku"             = try(local.anydb.os.sku, local.anydb_custom_image ? "" : "<sku>")
    "version"         = try(local.anydb.os.version, local.anydb_custom_image ? "" : "latest")
  }
```

File: deploy/terraform/modules/anydb_node/variables_local.tf (hunk 5 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf:18]: ok - reverting that

File: deploy/terraform/modules/anydb_node/variables_local.tf (hunk 6 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf:134]: fixed - now aligned with the hdb_node

File: deploy/terraform/modules/anydb_node/variables_local.tf (hunk 7 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf:136]: fixed

File: deploy/terraform/modules/anydb_node/variables_local.tf (hunk 8 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf:149]: fixed

File: deploy/terraform/modules/anydb_node/variables_local.tf (hunk 10 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf:103]: fixed like we spoke

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    "sku"             = try(local.anydb.os.sku, local.anydb_custom_image ? "" : local.os_defaults[upper(local.anydb_platform)].sku)
+    "version"         = try(local.anydb.os.version, local.anydb_custom_image ? "" : local.os_defaults[upper(local.anydb_platform)].version)
+  }
+
+  # Update database information with defaults
+  anydb_database = merge(local.anydb,
+    { platform = local.anydb_platform },
+    { db_version = local.anydb_version },
+    { size = local.anydb_size },
+    { os = local.anydb_ostype },
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:113]:
```suggestion
    { os = merge( {os_type = local.anydb_ostype}, local.anydb_os },
```

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+      "offer"     = "sles-sap-12-sp5",
+      "sku"       = "gen1"
+      "version"   = "latest"
+    }
+    HANA = {
+      "publisher" = "suse",
+      "offer"     = "sles-sap-12-sp5",
+      "sku"       = "gen1"
+      "version"   = "latest"
+    }
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:109]: HANA uses a different sizing and we have a separate module for that.
Let's not mix it in anydb for the time being.

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+      "offer"     = "SQL2017-WS2016",
+      "sku"       = "standard-gen2",
+      "version"   = "latest"
+    }
+    NONE = {
+      "publisher" = "suse",
+      "offer"     = "sles-sap-12-sp5",
+      "sku"       = "gen1"
+      "version"   = "latest"
+    }
```
Comment (rukawata) [inline deploy/terraform/modules/anydb_node/variables_local.tf:103]: NONE means no DB will be deployed, hence no os information is required.

File: deploy/terraform/modules/anydb_node/variables_local.tf (hunk 12 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf]: removed it

File: deploy/terraform/modules/anydb_node/variables_local.tf (hunk 13 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf]: removed it - allthough this will cause the automated tests to fail

File: deploy/terraform/modules/anydb_node/vm-anydb.tf [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+      caching              = disk.value.caching
+      storage_account_type = disk.value.disk_type
+      disk_size_gb         = disk.value.size_gb
+    }
+  }
+
+  computer_name                   = lower(format("%sxdbl%02d", upper(local.anydb_sid), count.index))
+  admin_username                  = local.authentication.username
+  disable_password_authentication = local.authentication.type != "password" ? true : false
+
```
Comment (YunziZhangAzure) [inline deploy/terraform/modules/anydb_node/vm-anydb.tf:57]: Missing: admin_password

File: deploy/terraform/modules/anydb_node/variables_local.tf [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+
+  anydb_ostype = try(local.anydb.os.os_type, "Linux")
+  anydb_size   = try(local.anydb.size, "500")
+  anydb_sku    = try(lookup(local.sizes, local.anydb_size).compute.vmsize, "Standard_E4s_v3")
+  anydb_fs     = try(local.anydb.filesystem, "xfs")
+  anydb_ha     = try(local.anydb.high_availability, "false")
+  anydb_sid    = (length(local.anydb-databases) > 0) ? try(local.anydb.instance.sid, "OR1") : "OR1"
+  loadbalancer = try(local.anydb.loadbalancer, {})
+
+  authentication = try(var.application.authentication,
```
Comment (YunziZhangAzure) [inline deploy/terraform/modules/anydb_node/variables_local.tf:63]: Why application? I think you might need to use authentication under database.

Comment (KimForss) [reply deploy/terraform/modules/anydb_node/variables_local.tf:63]: good catch,, fixed

File: deploy/terraform/modules/anydb_node/vm-anydb.tf (hunk 14 above)
Comment (KimForss) [reply deploy/terraform/modules/anydb_node/vm-anydb.tf:58]: added
